apps/server/lib/lexical/server/project/progress/state.ex

```diff
@@ -0,0 +1,88 @@
+defmodule Lexical.Server.Project.Progress.State do
+  alias Lexical.Project
+  alias Lexical.Protocol.Notifications
+  alias Lexical.Server.Project.Progress.Value
+  alias Lexical.Server.Transport
+
+  require Logger
+
+  defstruct project: nil, token_by_label: %{}, progress: nil
+
+  def new(%Project{} = project) do
+    %__MODULE__{project: project}
+  end
+
+  def begin(%__MODULE__{} = state, label) do
+    token = System.unique_integer([:positive])
+    token_by_label = Map.put(state.token_by_label, label, token)
+
+    state = %{state | token_by_label: token_by_label}
+    create_work_done(state, label)
+
+    progress = value(:begin, state, label, nil)
+    write(progress)
+
+    %{state | progress: progress}
+  end
+
+  def update(%__MODULE__{} = state, label, message) do
+    progress = value(:report, state, label, message)
+    write(progress)
+    %{state | progress: progress}
+  end
+
+  def complete(%__MODULE__{} = state, label, message) do
+    progress = value(:end, state, label, message)
+    write(progress)
+    state = clear_token_by_label(state, label)
+    %{state | progress: progress}
+  end
+
+  defp create_work_done(%__MODULE__{} = state, label) do
+    token = get_token(state, label)
+    progress = Notifications.CreateWorkDoneProgress.new(token: token)
+    Transport.write(progress)
+  end
+
+  defp write(progress) do
+    if progress && progress.token do
+      progress |> Value.to_progress() |> Transport.write()
+    end
+  end
+
+  defp value(:begin, state, label, _message) do
+    token = get_token(state, label)
+    Value.new(kind: :begin, token: token, title: label)
```

Comment:
heh, I might push this logic into value, so you have `begin`, `report` and `end` functions there

I'm a little wary of naming functions `end` though, how about `complete`  or `start` and `finish`?